ejb3/src/main/java/org/jboss/as/ejb3/component/EJBComponentDescription.java

```diff
@@ -1188,22 +1188,19 @@ public String toString() {
         }
 
         // Next interceptor: run-as-principal
-        // Switch users if there's a run-as principal
-        if (runAsPrincipal != null) {
-            interceptorFactories.put(InterceptorOrder.View.RUN_AS_PRINCIPAL, new ImmediateInterceptorFactory(new RunAsPrincipalInterceptor(runAsPrincipal)));
-
+        if (runAsPrincipal == null && !propagateSecurity) {
+            interceptorFactories.put(InterceptorOrder.View.RUN_AS_PRINCIPAL, new ImmediateInterceptorFactory(new RunAsPrincipalInterceptor(RunAsPrincipalInterceptor.ANONYMOUS_PRINCIPAL, false)));
+        } else {
+            // Switch users (set new incoming)
+            interceptorFactories.put(InterceptorOrder.View.RUN_AS_PRINCIPAL, new ImmediateInterceptorFactory(new RunAsPrincipalInterceptor(runAsPrincipal, isSecurityRequired())));
```

Comment:
@darranl We have to add RunAsPrincipalInterceptor to beans that do not have RunAsPrincipal annotation and propagate security. This is to be able to propagate incoming runas identity further. It should behave the same as legacy.